# IBM Applied D.S Capstone Project

## Camilo Echeverry

### Week 5

#### INDEX

In [1]:
import requests
import pandas as pd
import numpy as np
import folium

In [2]:
conda install -c anaconda xlrd --yes

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\DELL\anaconda3

  added / updated specs:
    - xlrd


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.10.14 |                0         159 KB  anaconda
    openssl-1.1.1h             |       he774522_0         5.8 MB  anaconda
    xlrd-1.2.0                 |             py_0         108 KB  anaconda
    ------------------------------------------------------------
                                           Total:         6.0 MB

The following packages will be SUPERSEDED by a higher-priority channel:

  ca-certificates                                 pkgs/main --> anaconda
  openssl                                       conda-forge --> anaconda
  xlrd                                            pkgs/main --> anaconda




openssl-1.1.1h       | 5.8 MB    |            |   0% 
op

In [3]:
conda install -c anaconda lxml --yes

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\DELL\anaconda3

  added / updated specs:
    - lxml


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    lxml-4.6.1                 |   py38h1350720_0         1.1 MB  anaconda
    ------------------------------------------------------------
                                           Total:         1.1 MB

The following packages will be SUPERSEDED by a higher-priority channel:

  lxml                                            pkgs/main --> anaconda




lxml-4.6.1           | 1.1 MB    |            |   0% 
lxml-4.6.1           | 1.1 MB    | 1          |   1% 
lxml-4.6.1           | 1.1 MB    | ##2        |  22% 
lxml-4.6.1           | 1.1 MB    | #####3     |  53% 
lxml-4.6.1           | 1.1 MB    | ########2  |  83% 
lxml-4.6.1           | 1.1 MB    | ########## | 100% 
Preparing transaction: ...wo

In [8]:
url  = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
page = requests.get(url)
if page.status_code == 200:
    print('Page download successful')
else:
    print('Page download error. Error code: {}'.format(page.status_code))

Page download successful


In [9]:
import lxml
df_html = pd.read_html(url, header=0, na_values = ['Not assigned'])[0]
df_html.head()

Postal Code           Borough              Neighbourhood
0         M1A               NaN                        NaN
1         M2A               NaN                        NaN
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [10]:
df_html.dropna(subset=['Borough'], inplace=True)

In [14]:
n_empty_neighborhood = df_html[df_html['Neighbourhood'].isna()].shape[0]
print('Number of rows on which Neighborhood column is empty: {}'.format(n_empty_neighborhood))

Number of rows on which Neighborhood column is empty: 0


In [15]:
df_html[df_html['Neighbourhood'].isna()]

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

In [16]:
df_html['Neighbourhood'].fillna(df_html['Borough'], inplace=True)
n_empty_neighborhood = df_html[df_html['Neighbourhood'].isna()].shape[0]
print('Number of rows on which Neighborhood column is empty: {}'.format(n_empty_neighborhood))

Number of rows on which Neighborhood column is empty: 0


In [21]:
df_html[df_html['Borough']=="Queen's Park"]

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

In [23]:
df_postcodes =pd.DataFrame( df_html.groupby(['Postal Code','Borough']).Neighbourhood.agg([('Neighbourhood', ', '.join)]))
df_postcodes.reset_index(inplace=True)
df_postcodes.head(5)

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [24]:
df_html.dropna(subset=['Borough'], inplace=True)

In [25]:
n_empty_neighborhood = df_html[df_html['Neighbourhood'].isna()].shape[0]
print('Number of rows on which Neighborhood column is empty: {}'.format(n_empty_neighborhood))

Number of rows on which Neighborhood column is empty: 0


In [26]:
print('The shape of the dataset is:',df_postcodes.shape)

The shape of the dataset is: (103, 3)


In [27]:
df_postcodes.to_csv('Toronto_Postcodes.csv')

In [28]:
url_csv = 'http://cocl.us/Geospatial_data'
df_coordinates = pd.read_csv(url_csv)
df_coordinates.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [29]:
df_neighborhoods = pd.read_csv('Toronto_Postcodes.csv',index_col=[0])
df_neighborhoods.head()

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [33]:
df_coordinates.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)
df_neighborhoods.rename(columns={'Postal Code': 'PostalCode'}, inplace=True)

In [34]:
df_neighborhoods_coordinates = pd.merge(df_neighborhoods, df_coordinates, on='PostalCode')
df_neighborhoods_coordinates.head()

PostalCode      Borough                           Neighbourhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [35]:
df_neighborhoods_coordinates[(df_neighborhoods_coordinates['PostalCode']=='M5G') |
                             (df_neighborhoods_coordinates['PostalCode']=='M2H') ]

PostalCode           Borough       Neighbourhood   Latitude  Longitude
17        M2H        North York   Hillcrest Village  43.803762 -79.363452
57        M5G  Downtown Toronto  Central Bay Street  43.657952 -79.387383

In [36]:
df_neighborhoods_coordinates.to_csv('Toronto_Postcodes_2.csv')

In [37]:
df = pd.read_csv('Toronto_Postcodes_2.csv', index_col=0)
df.head()

PostalCode      Borough                           Neighbourhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [38]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df['Borough'].unique()),
        df.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


In [39]:
df.rename(columns={'Neighbourhood': 'Neighborhood'}, inplace=True)

In [40]:
df.groupby('Borough').count()['Neighborhood']

Borough
Central Toronto      9
Downtown Toronto    19
East Toronto         5
East York            5
Etobicoke           12
Mississauga          1
North York          24
Scarborough         17
West Toronto         6
York                 5
Name: Neighborhood, dtype: int64

In [41]:
df_toronto = df[df['Borough'].str.contains('Toronto')]
df_toronto.reset_index(inplace=True)
df_toronto.drop('index', axis=1, inplace=True)
df_toronto.head()

C:\Users\DELL\anaconda3\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


PostalCode          Borough                    Neighborhood   Latitude  \
0        M4E     East Toronto                     The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
3        M4M     East Toronto                 Studio District  43.659526   
4        M4N  Central Toronto                   Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [42]:
print(df_toronto.groupby('Borough').count()['Neighborhood'])

Borough
Central Toronto      9
Downtown Toronto    19
East Toronto         5
West Toronto         6
Name: Neighborhood, dtype: int64


In [43]:
boroughs = df_toronto['Borough'].unique().tolist()

In [44]:
lat_toronto = df_toronto['Latitude'].mean()
lon_toronto = df_toronto['Longitude'].mean()
print('The geographical coordinates of Toronto are {}, {}'.format(lat_toronto, lon_toronto))

The geographical coordinates of Toronto are 43.66713498717948, -79.38987324871795


In [45]:
borough_color = {}
for borough in boroughs:
    borough_color[borough]= '#%02X%02X%02X' % tuple(np.random.choice(range(256), size=3))

In [46]:
map_toronto = folium.Map(location=[lat_toronto, lon_toronto], zoom_start=12)

for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], 
                                           df_toronto['Longitude'],
                                           df_toronto['Borough'], 
                                           df_toronto['Neighborhood']):
    label_text = borough + ' - ' + neighborhood
    label = folium.Popup(label_text)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color=borough_color[borough],
        fill_color=borough_color[borough],
        fill_opacity=0.7).add_to(map_toronto)  
    
map_toronto

In [52]:
CLIENT_ID = 'XUBKBKKLLCKFRHBUTVACDNQZN1EJWZLMSRJJG05AEM3KHY50'
CLIENT_SECRET = 'SYQMGEAXIG3TONPXUAU3SBWFIICNB4QEJBXFMHNF1L2JZCFI'
VERSION = '20180605'
LIMIT = 100
radius = 500

In [53]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
   
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [54]:
toronto_venues = getNearbyVenues(names=df_toronto['Neighborhood'],
                                latitudes=df_toronto['Latitude'],
                                longitudes=df_toronto['Longitude'])

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West, Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High 

In [55]:
toronto_venues.shape

(1615, 7)

In [56]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing Centre, South Ce...                     16   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     60   
Christie                                                               15   
Church and Wellesley                                                   79   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                       10   
Dufferin, Dovercourt Village                                           17   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              63   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               43   
Moore Park, Summerhill East                                             4   
North Toronto West, Lawrence Park                                      20   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            37   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               96   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     36   
St. James Town                                                         79   
St. James Town, Cabbagetown                                            47   
Stn A PO Boxes                                                         98   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     15   
The Annex, North Midtown, Yorkville                                    20   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         32   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
Brockton, Parkdale Village, Exhibition Place                            22   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      60   
Christie                                                                15   
Church and Wellesley                                                    79   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              33   
Davisville North                    

In [57]:
toronto_venues['Venue Category'].unique()[:100]

array(['Trail', 'Health Food Store', 'Pub', 'Neighborhood',
       'Cosmetics Shop', 'Greek Restaurant', 'Italian Restaurant',
       'Ice Cream Shop', 'Yoga Studio', 'Fruit & Vegetable Store',
       'Brewery', 'Restaurant', 'Pizza Place', 'Juice Bar', 'Bookstore',
       'Furniture / Home Store', 'Dessert Shop', 'Bubble Tea Shop',
       'Grocery Store', 'Spa', 'Coffee Shop', 'Tibetan Restaurant',
       'Bakery', 'Indian Restaurant', 'Caribbean Restaurant', 'Café',
       'Japanese Restaurant', 'American Restaurant', 'Lounge',
       'Frozen Yogurt Shop', 'Liquor Store', 'Gym', 'Fish & Chips Shop',
       'Fast Food Restaurant', 'Sushi Restaurant', 'Park', 'Pet Store',
       'Steakhouse', 'Burrito Place', 'Movie Theater', 'Sandwich Place',
       'Fish Market', 'Seafood Restaurant', 'Gay Bar', 'Cheese Shop',
       'Comfort Food Restaurant', 'Middle Eastern Restaurant',
       'Stationery Store', 'Thai Restaurant', 'Coworking Space',
       'Gastropub', 'Clothing Store', 'Gym / Fit

In [58]:
"Indian Restaurant" in toronto_venues['Venue Category'].unique()

True

In [59]:
to_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
to_onehot['Neighborhoods'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [to_onehot.columns[-1]] + list(to_onehot.columns[:-1])
to_onehot = to_onehot[fixed_columns]

print(to_onehot.shape)
to_onehot.head()

(1615, 236)


Neighborhoods  Airport  Airport Food Court  Airport Gate  \
0                   The Beaches        0                   0             0   
1                   The Beaches        0                   0             0   
2                   The Beaches        0                   0             0   
3                   The Beaches        0                   0             0   
4  The Danforth West, Riverdale        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Theme Restaurant  Tibetan Restaurant  \
0             0         0  ...                 0                   0   
1             0         0  ...                 0                   0   
2             0         0  ...                 0                   0   
3             0         0  ...                 0                   0   
4             0         0  ...                 0                   0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      1              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Yoga Studio  
0                 0                      0         0            0  
1                 0                      0         0            0  
2                 0                      0         0            0  
3                 0                      0         0            0  
4                 0                      0         0            0  

[5 rows x 236 columns]

In [60]:
to_grouped = to_onehot.groupby(["Neighborhoods"]).mean().reset_index()

print(to_grouped.shape)
to_grouped

(39, 236)


Neighborhoods   Airport  \
0                                         Berczy Park  0.000000   
1        Brockton, Parkdale Village, Exhibition Place  0.000000   
2   Business reply mail Processing Centre, South C...  0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...  0.066667   
4                                  Central Bay Street  0.000000   
5                                            Christie  0.000000   
6                                Church and Wellesley  0.000000   
7                      Commerce Court, Victoria Hotel  0.000000   
8                                          Davisville  0.000000   
9                                    Davisville North  0.000000   
10                       Dufferin, Dovercourt Village  0.000000   
11             First Canadian Place, Underground city  0.000000   
12    Forest Hill North & West, Forest Hill Road Park  0.000000   
13                           Garden District, Ryerson  0.000000   
14  Harbourfront East, Union Station, Toronto Islands  0.000000   
15                      High Park, The Junction South  0.000000   
16                     India Bazaar, The Beaches West  0.000000   
17          Kensington Market, Chinatown, Grange Park  0.000000   
18                                      Lawrence Park  0.000000   
19                           Little Portugal, Trinity  0.000000   
20                        Moore Park, Summerhill East  0.000000   
21                  North Toronto West, Lawrence Park  0.000000   
22                             Parkdale, Roncesvalles  0.000000   
23        Queen's Park, Ontario Provincial Government  0.000000   
24                          Regent Park, Harbourfront  0.000000   
25                           Richmond, Adelaide, King  0.000000   
26                                           Rosedale  0.000000   
27                                           Roselawn  0.000000   
28                                 Runnymede, Swansea  0.000000   
29                                     St. James Town  0.000000   
30                        St. James Town, Cabbagetown  0.000000   
31                                     Stn A PO Boxes  0.000000   
32                                    Studio District  0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...  0.000000   
34                The Annex, North Midtown, Yorkville  0.000000   
35                                        The Beaches  0.000000   
36                       The Danforth West, Riverdale  0.000000   
37           Toronto Dominion Centre, Design Exchange  0.000000   
38                     University of Toronto, Harbord  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.000000      0.000000        0.000000         0.000000   
3             0.066667      0.066667        0.133333         0.066667   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000        0.000000         0.000000   
11            0.000000      0.000000        0.000000         0.000000   
12            0.000000      0.000000        0.000000         0.000000   
13            0.000000      0.000000        0.000000         0.000000   
14            0.000000      0.000000        0.000000         0.000000   
15            0.000000      0.000000        0.000000         0.000000   
16            0.000000      0.000000        0.000000         0.000000   
17            0.000000      0.000000        0

In [61]:
len(to_grouped[to_grouped["Indian Restaurant"] > 0])

7

In [62]:
to_indian = to_grouped[["Neighborhoods","Indian Restaurant"]]

In [63]:
to_indian.head(9)

Neighborhoods  Indian Restaurant
0                                        Berczy Park           0.000000
1       Brockton, Parkdale Village, Exhibition Place           0.000000
2  Business reply mail Processing Centre, South C...           0.000000
3  CN Tower, King and Spadina, Railway Lands, Har...           0.000000
4                                 Central Bay Street           0.016667
5                                           Christie           0.000000
6                               Church and Wellesley           0.012658
7                     Commerce Court, Victoria Hotel           0.000000
8                                         Davisville           0.030303

In [66]:
from sklearn.cluster import KMeans
toclusters = 3

to_clustering = to_indian.drop(["Neighborhoods"], 1)

kmeans = KMeans(n_clusters=toclusters, random_state=1)
kmeans.fit_transform(to_clustering)

kmeans.labels_[0:20]

array([0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [67]:
to_merged = to_indian.copy()

to_merged["Cluster Labels"] = kmeans.labels_

In [68]:
to_merged.rename(columns={"Neighborhoods": "Neighborhood"}, inplace=True)
to_merged.head(5)

Neighborhood  Indian Restaurant  \
0                                        Berczy Park           0.000000   
1       Brockton, Parkdale Village, Exhibition Place           0.000000   
2  Business reply mail Processing Centre, South C...           0.000000   
3  CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
4                                 Central Bay Street           0.016667   

   Cluster Labels  
0               0  
1               0  
2               0  
3               0  
4               1

In [69]:
to_merged = to_merged.join(toronto_venues.set_index("Neighborhood"), on="Neighborhood")

print(to_merged.shape)
to_merged.head()

(1615, 9)


Neighborhood  Indian Restaurant  Cluster Labels  Neighborhood Latitude  \
0  Berczy Park                0.0               0              43.644771   
0  Berczy Park                0.0               0              43.644771   
0  Berczy Park                0.0               0              43.644771   
0  Berczy Park                0.0               0              43.644771   
0  Berczy Park                0.0               0              43.644771   

   Neighborhood Longitude                                 Venue  \
0              -79.373306  The Keg Steakhouse + Bar - Esplanade   
0              -79.373306                                  LCBO   
0              -79.373306                        Fresh On Front   
0              -79.373306                         Meridian Hall   
0              -79.373306                           Berczy Park   

   Venue Latitude  Venue Longitude                 Venue Category  
0       43.646712       -79.374768                     Restaurant  
0       43.642944       -79.372440                   Liquor Store  
0       43.647815       -79.374453  Vegetarian / Vegan Restaurant  
0       43.646292       -79.376022                   Concert Hall  
0       43.648048       -79.375172                           Park

In [70]:
to_merged.sort_values(["Cluster Labels"], inplace=True)
to_merged.head()

Neighborhood  Indian Restaurant  Cluster Labels  \
0                Berczy Park                0.0               0   
25  Richmond, Adelaide, King                0.0               0   
25  Richmond, Adelaide, King                0.0               0   
25  Richmond, Adelaide, King                0.0               0   
25  Richmond, Adelaide, King                0.0               0   

    Neighborhood Latitude  Neighborhood Longitude  \
0               43.644771              -79.373306   
25              43.650571              -79.384568   
25              43.650571              -79.384568   
25              43.650571              -79.384568   
25              43.650571              -79.384568   

                                     Venue  Venue Latitude  Venue Longitude  \
0     The Keg Steakhouse + Bar - Esplanade       43.646712       -79.374768   
25                      Soho House Toronto       43.648734       -79.386541   
25  The Keg Steakhouse + Bar - York Street       43.649987       -79.384103   
25                               Rosalinda       43.650252       -79.385156   
25                      Shangri-La Toronto       43.649129       -79.386557   

                   Venue Category  
0                      Restaurant  
25                      Speakeasy  
25                     Restaurant  
25  Vegetarian / Vegan Restaurant  
25                          Hotel

In [71]:
map_clusters = folium.Map(location=[lat_toronto, lon_toronto],zoom_start=14)

markers_colors={}
markers_colors[0] = 'red'
markers_colors[1] = 'blue'
markers_colors[2] = 'green'
markers_colors[3] = 'yellow'
markers_colors[4] = 'cyan'
markers_colors[5] = 'black'
for lat, lon, cluster in zip(to_merged['Neighborhood Latitude'], to_merged['Neighborhood Longitude'], to_merged['Cluster Labels']):
    
    
    folium.features.CircleMarker(
        [lat, lon],
        radius=5,
       
        color =markers_colors[cluster],
        fill_color=markers_colors[cluster],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [72]:
to_merged.loc[(to_merged['Cluster Labels'] ==0) & (to_merged['Venue Category'] == 'Indian Restaurant') ]

Neighborhood  Indian Restaurant  \
14  Harbourfront East, Union Station, Toronto Islands               0.01   

    Cluster Labels  Neighborhood Latitude  Neighborhood Longitude  \
14               0              43.640816              -79.381752   

                Venue  Venue Latitude  Venue Longitude     Venue Category  
14  Indian Roti House        43.63906       -79.385422  Indian Restaurant

In [73]:
to_merged.loc[(to_merged['Cluster Labels'] ==1) & (to_merged['Venue Category'] == 'Indian Restaurant') ]

Neighborhood  Indian Restaurant  Cluster Labels  \
36  The Danforth West, Riverdale           0.023256               1   
8                     Davisville           0.030303               1   
30   St. James Town, Cabbagetown           0.021277               1   
4             Central Bay Street           0.016667               1   
6           Church and Wellesley           0.012658               1   

    Neighborhood Latitude  Neighborhood Longitude                   Venue  \
36              43.679557              -79.352188           Sher-E-Punjab   
8               43.704324              -79.388790  Marigold Indian Bistro   
30              43.667967              -79.367675  Butter Chicken Factory   
4               43.657952              -79.387383         Colaba Junction   
6               43.665860              -79.383160   Kothur Indian Cuisine   

    Venue Latitude  Venue Longitude     Venue Category  
36       43.677308       -79.353066  Indian Restaurant  
8        43.702881       -79.388008  Indian Restaurant  
30       43.667072       -79.369184  Indian Restaurant  
4        43.660940       -79.385635  Indian Restaurant  
6        43.667872       -79.385659  Indian Restaurant

In [74]:
to_merged.loc[(to_merged['Cluster Labels'] ==2) & (to_merged['Venue Category'] == 'Indian Restaurant') ]

Neighborhood  Indian Restaurant  Cluster Labels  \
34  The Annex, North Midtown, Yorkville               0.05               2   

    Neighborhood Latitude  Neighborhood Longitude                  Venue  \
34               43.67271              -79.405678  Roti Cuisine of India   

    Venue Latitude  Venue Longitude     Venue Category  
34       43.674618       -79.408249  Indian Restaurant

#### Insights:

Most of the Indian restaurants are in cluster 2 which is around
And lowest in Cluster 1 areas which are in North Toronto West and Parkade areas.Also, there are good opportunities to open
near St James Town, Cabbagetown, Toronto Centre, Regent Park, Chruch Wellesley Village.
Looking at nearby venues it seems cluster 0 might be a good location as there are not a lot of Indian restaurants in these areas. Therefore, this project recommends the entrepreneur to open an authentic Indian restaurant in these locations.
Furthermore, sing k-means cluster algorithm I separated the neighborhood into 10(Ten) different clusters and for 103 different lattitude and logitude from dataset, which have very-similar neighborhoods around them. Using the charts above results presented to a particular neighborhood based on average house prices and school rating have been made

##### Notebook Finished